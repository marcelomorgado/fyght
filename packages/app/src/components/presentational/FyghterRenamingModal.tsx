import React, { useState } from "react";
import { Button, Modal, Form, Input } from "antd";
import { useFyghtContext } from "../../context";

// interface Values {}

interface FyghterRenamingFormProps {
  visible: boolean;
  // TODO: To use Values type above
  // See more: https://github.com/ant-design/ant-design/issues/21195
  onSave: (values: {}) => void;
  onCancel: () => void;
  name: string;
}

const FyghterRenamingForm: React.FC<FyghterRenamingFormProps> = ({ visible, onCancel, onSave, name }) => {
  const [form] = Form.useForm();
  return (
    <Modal
      visible={visible}
      title="Rename fyghter"
      okText="Save"
      onCancel={onCancel}
      onOk={async (): Promise<void> => {
        const values = await form.validateFields();
        form.resetFields();
        onSave(values);
      }}
    >
      <Form
        form={form}
        layout="vertical"
        name="form_in_modal"
        initialValues={{
          name,
        }}
      >
        <Form.Item
          name="name"
          label="Name"
          rules={[
            {
              required: true,
              message: "Please input the name of the fyghter!",
            },
          ]}
        >
          <Input />
        </Form.Item>
      </Form>
    </Modal>
  );
};

export const FyghterRenamingModal: React.FC = () => {
  const [isVisible, setVisible] = useState(false);

  const {
    renameMyFyghter,
    setErrorMessage,
    state: { myFyghter },
  } = useFyghtContext();

  const onSave = async ({ name }: { name: string }): Promise<void> => {
    try {
      renameMyFyghter(name);
      setVisible(false);
    } catch (e) {
      setErrorMessage("Unexpected error when creating Fyghter.");
    }
  };

  const isLoading = !myFyghter || !myFyghter.id ? true : false;

  return (
    <div>
      <Button
        type="primary"
        block={true}
        onClick={(): void => {
          setVisible(true);
        }}
        loading={isLoading}
      >
        Rename
      </Button>
      <FyghterRenamingForm
        name={myFyghter.name}
        visible={isVisible}
        onCancel={(): void => {
          setVisible(false);
        }}
        onSave={onSave}
      />
    </div>
  );
};
