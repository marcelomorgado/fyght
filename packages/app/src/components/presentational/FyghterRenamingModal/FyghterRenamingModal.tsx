import React, { useState } from "react";
import { Button, Modal, Form, Input } from "antd";
import { useFyghtState } from "../../../state";

// interface Values {}

interface FyghterRenamingFormProps {
  visible: boolean;
  // TODO: To use Values type above
  // See more: https://github.com/ant-design/ant-design/issues/21195
  onSave: (values: {}) => void;
  onCancel: () => void;
  name: string;
}

const FyghterRenamingForm: React.FC<FyghterRenamingFormProps> = ({ visible, onCancel, onSave, name }) => {
  const [form] = Form.useForm();
  return (
    <Modal
      visible={visible}
      title="Rename fyghter"
      okText="Save"
      onCancel={onCancel}
      onOk={async (): Promise<void> => {
        const values = await form.validateFields();
        form.resetFields();
        onSave(values);
      }}
    >
      <Form
        form={form}
        layout="vertical"
        name="form_in_modal"
        initialValues={{
          name,
        }}
      >
        <Form.Item
          name="name"
          label="Name"
          rules={[
            {
              required: true,
              message: "Please input the name of the fyghter!",
            },
          ]}
        >
          <Input />
        </Form.Item>
      </Form>
    </Modal>
  );
};

type Props = {
  isLoading: boolean;
};

// TODO: Rename component
export const FyghterRenamingModal: React.FC<Props> = ({ isLoading }: Props) => {
  const [isVisible, setVisible] = useState(false);

  const [{ myFyghter }, { renameMyFyghter }] = useFyghtState();

  const onSave = async ({ name }: { name: string }): Promise<void> => {
    renameMyFyghter(name);
    setVisible(false);
  };

  return (
    <div>
      <Button
        type="primary"
        block={true}
        onClick={(): void => {
          setVisible(true);
        }}
        loading={isLoading}
      >
        Rename
      </Button>
      <FyghterRenamingForm
        name={myFyghter.name}
        visible={isVisible}
        onCancel={(): void => {
          setVisible(false);
        }}
        onSave={onSave}
      />
    </div>
  );
};
