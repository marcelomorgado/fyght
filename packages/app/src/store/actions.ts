import {
  ethers,
  ContractTransaction,
  Event,
  Transaction,
  ContractReceipt,
  Contract,
} from "ethers";
import { BigNumber } from "ethers";

const { getAddress } = ethers.utils;

export const RENAME = "RENAME";
export const CHANGE_SKIN = "CHANGE_SKIN";
export const INCREMENT_MY_FIGHTER_XP = "INCREMENT_MY_FIGHTER_XP";
export const INCREMENT_ENEMY_XP = "INCREMENT_ENEMY_XP";
export const LOAD_ENEMIES = "LOAD_ENEMIES";
export const SET_MY_FYGHTER = "SET_MY_FYGHTER";
export const UPDATE_METAMASK_ACCOUNT = "UPDATE_METAMASK_ACCOUNT";
export const UPDATE_METAMASK_NETWORK = "UPDATE_METAMASK_NETWORK";
export const INITIALIZE_METAMASK = "INITIALIZE_METAMASK";

interface NewFyghter {
  owner: string;
  id: BigNumber;
  name: string;
}

export const createActions = (dispatch: any, state: FyghtContext): any => {
  const incrementMyFyghterXp = (): void =>
    dispatch({ type: INCREMENT_MY_FIGHTER_XP, payload: {} });

  const incrementEnemyXp = (enemyId: BigNumber): void =>
    dispatch({ type: INCREMENT_ENEMY_XP, payload: { enemyId } });

  const setEnemies = (enemies: Fyghter[]): void =>
    dispatch({ type: LOAD_ENEMIES, payload: { enemies } });

  const setMyFyghter = (myFyghter: Fyghter): void =>
    dispatch({ type: SET_MY_FYGHTER, payload: { myFyghter } });

  const createFyghter = async (name: string): Promise<void> => {
    const {
      metamask: { contract: fyghters },
    } = state;

    try {
      const tx: ContractTransaction = await fyghters.create(name);
      await tx.wait();
      console.log(tx);

      // TODO: Get event from transaction
      const filter = fyghters.filters.NewFyghter(null, null, null);
      fyghters.on(filter, async (owner: string, id: number, name: string) => {
        const myFyghter = await fyghters.fyghters(id);
        setMyFyghter(myFyghter);
      });
    } catch (e) {
      console.log(e);
      // Revert message
      //console.log(e.data.message);
    }
  };

  const optimisticUpdate = async (
    tx: Transaction,
    onSuccess: () => void,
    onError: () => void
  ): Promise<void> => {
    const {
      metamask: { provider },
    } = state;

    // Optimistic update
    onSuccess();

    provider.once(tx.hash, (receipt: ContractReceipt) => {
      const { status } = receipt;
      if (!status) {
        onError();
      }
    });
  };

  const renameMyFyghter = async (newName: string): void => {
    const {
      myFyghter,
      metamask: { contract: fyghters },
    } = state;
    const { id: myFyghterId, name: oldName } = myFyghter;

    const tx: ContractTransaction = await fyghters.rename(myFyghterId, newName);

    optimisticUpdate(
      tx,
      () => {
        dispatch({ type: RENAME, payload: { name: newName } });
      },
      () => {
        dispatch({ type: RENAME, payload: { name: oldName } });
      }
    );
  };

  const changeMyFyghterSkin = async (newSkin: string): Promise<void> => {
    const {
      myFyghter,
      metamask: { contract: fyghters },
    } = state;
    const { id: myFyghterId, skin: oldSkin } = myFyghter;

    const tx: ContractTransaction = await fyghters.changeSkin(
      myFyghterId,
      newSkin
    );

    optimisticUpdate(
      tx,
      () => {
        dispatch({ type: CHANGE_SKIN, payload: { skin: newSkin } });
      },
      () => {
        dispatch({ type: RENAME, payload: { name: oldSkin } });
      }
    );
  };

  const attackAnEnemy = async (enemyId: BigNumber): Promise<void> => {
    const {
      myFyghter: { id: myFyghterId },
      metamask: { contract: fyghters, provider },
    } = state;

    const tx: ContractTransaction = await fyghters.attack(myFyghterId, enemyId);

    provider.once(tx.hash, ({ status }: ContractReceipt) => {
      if (!status) {
        // TODO: Error message
      }
    });

    const filter = fyghters.filters.Attack(myFyghterId, null, null);
    fyghters.once(
      filter,
      async (
        myFighterId: BigNumber,
        enemyId: BigNumber,
        winnerId: BigNumber
      ) => {
        if (winnerId.eq(myFighterId)) {
          incrementMyFyghterXp();
        } else {
          incrementEnemyXp(enemyId);
        }
      }
    );
  };

  //
  // Note: This low level code needed to get past events is the way that ethers.js v4 works
  // See more: https://github.com/marcelomorgado/fyght/issues/78
  //
  const loadEnemies = async (): Promise<void> => {
    const {
      metamask: { contract: fyghters, provider, account },
    } = state;

    if (!fyghters || !provider) {
      setEnemies([]);
      return;
    }

    const filter = fyghters.filters.NewFyghter(null, null, null);
    const logs = await fyghters.queryFilter(filter, 0, "latest");
    const enemiesIds = logs
      .map((l: Event) => l.args)
      .filter(
        ({ owner }: NewFyghter) => getAddress(owner) !== getAddress(account)
      )
      .map(({ id }: NewFyghter) => id);

    const enemiesPromises = enemiesIds.map((id: BigNumber) =>
      fyghters.fyghters(id)
    );
    const enemies: Fyghter[] = await Promise.all(enemiesPromises);

    setEnemies(enemies);
  };

  const loadMyFyghter = async (): Promise<void> => {
    const {
      metamask: { contract: fyghters, account },
    } = state;

    const filter = fyghters.filters.NewFyghter(getAddress(account), null, null);
    const logs = await fyghters.queryFilter(filter, 0, "latest");
    const [myFyghterId] = logs
      .map((l: Event) => l.args)
      .map(({ id }: NewFyghter) => id);

    if (myFyghterId) {
      const myFyghter = await fyghters.fyghters(myFyghterId);
      setMyFyghter(myFyghter);
    }
  };

  const setMetamaskAccount = (account: string): void =>
    dispatch({ type: UPDATE_METAMASK_ACCOUNT, payload: { account } });

  const setMetamaskNetworkId = (networkId: number): void =>
    dispatch({ type: UPDATE_METAMASK_NETWORK, payload: { networkId } });

  const initializeMetamask = (): void => {
    const {
      metamask: { ethereum },
    } = state;

    if (ethereum) {
      // Note: The metamask docs recommends to use the 'chainChanged' event instead but it isn't working
      // See more: https://docs.metamask.io/guide/ethereum-provider.html#methods-new-api
      ethereum.on("networkChanged", (networkId: number) => {
        setMetamaskNetworkId(networkId);
      });

      ethereum.on("accountsChanged", ([account]: string[]) => {
        setMetamaskAccount(account);
      });
    }

    dispatch({ type: INITIALIZE_METAMASK, payload: {} });
  };

  return {
    renameMyFyghter,
    changeMyFyghterSkin,
    attackAnEnemy,
    loadEnemies,
    loadMyFyghter,
    setMetamaskAccount,
    setMetamaskNetworkId,
    initializeMetamask,
    createFyghter,
  };
};
