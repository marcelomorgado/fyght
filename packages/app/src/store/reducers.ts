import { ethers } from "ethers";
import { FyghtersFactory } from "../contracts/FyghtersFactory";
import { Fyghters } from "../contracts/Fyghters";
import {
  RENAME,
  CHANGE_SKIN,
  UPDATE_ENEMY_XP,
  UPDATE_MY_FIGHTER_XP,
  LOAD_ENEMIES,
  SET_MY_FYGHTER,
  UPDATE_METAMASK_ACCOUNT,
  UPDATE_METAMASK_NETWORK,
  INITIALIZE_METAMASK,
} from "./actions";
import { Metamask, Fyghter, Action, FyghtContextInterface } from "../global";

declare global {
  interface Window {
    // TODO: Set properly type
    ethereum: any;
  }
}

const { ethereum } = window;
ethereum.autoRefreshOnNetworkChange = false;

export const initialState: FyghtContextInterface = {
  myFyghter: null,
  enemies: [],
  metamask: {
    networkId: ethereum.networkVersion,
    account: null,
    ethereum,
    contract: null,
    provider: null,
  },
};

const myFyghterReducer = (
  state: Fyghter = initialState.myFyghter,
  action: Action
): Fyghter => {
  const { type, payload } = action;
  const { name, skin, xp, myFyghter } = payload;

  switch (type) {
    case RENAME:
      return { ...state, name };
    case CHANGE_SKIN:
      return { ...state, skin };
    case UPDATE_MY_FIGHTER_XP:
      return { ...state, xp };
    case SET_MY_FYGHTER:
      return myFyghter;
    default:
      return state;
  }
};

const enemiesReducer = (
  state: Array<Fyghter> = initialState.enemies,
  action: Action
): Array<Fyghter> => {
  const { type, payload } = action;
  const { enemyId, xp, enemies } = payload;

  switch (type) {
    case UPDATE_ENEMY_XP:
      return state.map(e => {
        if (e.id === enemyId) {
          return { ...e, xp };
        } else {
          return e;
        }
      });
    case LOAD_ENEMIES:
      return enemies;
    default:
      return state;
  }
};

const metamaskReducer = (
  state: Metamask = initialState.metamask,
  action: Action
): Metamask => {
  const { type, payload } = action;
  const { networkId, account } = payload;

  switch (type) {
    case UPDATE_METAMASK_ACCOUNT:
      return { ...state, account };
    case UPDATE_METAMASK_NETWORK:
      return { ...state, networkId };
    case INITIALIZE_METAMASK: {
      const { ethereum } = state;
      // TODO: Move to a setup/env config
      const FYGHTERS_CONTRACT_ADDRESS =
        "0x49de9b5f6c0Dc3e22e9Af986477Cac01dBe82659";
      const provider = new ethers.providers.Web3Provider(ethereum);
      const contract: Fyghters = FyghtersFactory.connect(
        FYGHTERS_CONTRACT_ADDRESS,
        // TODO: Change signer when accounts change
        provider.getSigner()
      );

      return { ...state, contract, provider, ethereum };
    }
    default:
      return state;
  }
};

export const rootReducer = (
  state: FyghtContextInterface = initialState,
  action: Action
): FyghtContextInterface => {
  const { myFyghter, enemies, metamask } = state;
  return {
    myFyghter: myFyghterReducer(myFyghter, action),
    enemies: enemiesReducer(enemies, action),
    metamask: metamaskReducer(metamask, action),
  };
};
