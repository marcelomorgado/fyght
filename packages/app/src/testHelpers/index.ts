import { BigNumber } from "ethers";
import { Skin } from "../constants";

export const storeMocks: { myFyghter: Fyghter; enemies: Enemy[] } = {
  myFyghter: {
    id: BigNumber.from("1"),
    name: "John",
    skin: Skin.NORMAL_GUY,
    xp: BigNumber.from("5"),
    balance: BigNumber.from("5"),
  },
  enemies: [
    {
      winProbability: BigNumber.from("50"),
      fyghter: {
        id: BigNumber.from("2"),
        name: "Charlie",
        skin: Skin.NAKED,
        xp: BigNumber.from("1"),
        balance: BigNumber.from("5"),
      },
    },
    {
      winProbability: BigNumber.from("50"),
      fyghter: {
        id: BigNumber.from("3"),
        name: "Saul",
        skin: Skin.NINJA,
        xp: BigNumber.from("10"),
        balance: BigNumber.from("5"),
      },
    },
    {
      winProbability: BigNumber.from("50"),
      fyghter: {
        id: BigNumber.from("4"),
        name: "Julian",
        skin: Skin.NAKED,
        xp: BigNumber.from("1"),
        balance: BigNumber.from("5"),
      },
    },
    {
      winProbability: BigNumber.from("50"),
      fyghter: {
        id: BigNumber.from("5"),
        name: "Brad",
        skin: Skin.MASTER,
        xp: BigNumber.from("89"),
        balance: BigNumber.from("5"),
      },
    },
    {
      winProbability: BigNumber.from("50"),
      fyghter: {
        id: BigNumber.from("6"),
        name: "Frank",
        skin: Skin.NO_ONE,
        xp: BigNumber.from("71"),
        balance: BigNumber.from("5"),
      },
    },
    {
      winProbability: BigNumber.from("50"),
      fyghter: {
        id: BigNumber.from("7"),
        name: "Brian",
        skin: Skin.MONK,
        xp: BigNumber.from("30"),
        balance: BigNumber.from("5"),
      },
    },
    {
      winProbability: BigNumber.from("50"),
      fyghter: {
        id: BigNumber.from("8"),
        name: "Fred",
        skin: Skin.KARATE_KID,
        xp: BigNumber.from("10"),
        balance: BigNumber.from("5"),
      },
    },
  ],
};
