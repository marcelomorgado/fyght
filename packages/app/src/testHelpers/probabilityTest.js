const p = (myXp, hisXp) => 1000 / (myXp + hisXp);

for (let myXp = 1; myXp <= 100; ++myXp) {
  for (let hisXp = 1; hisXp <= 100; ++hisXp) {
    console.log(`myXp=${myXp}, hisXp=${hisXp}, p=${p(myXp, hisXp)}`);
  }
}
