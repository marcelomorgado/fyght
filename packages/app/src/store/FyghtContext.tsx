import React, { useReducer, useContext } from "react";
import { createActions } from "./actions";
import { rootReducer, initialState } from "./reducers";

const FyghtContext = React.createContext<any | null>(null);

const asyncer = (dispatch: any, state: any) => (action: any) =>
  typeof action === "function" ? action(dispatch, state) : dispatch(action);

const FyghtProvider = ({ children }: any) => {
  const [state, baseDispatch] = useReducer(rootReducer, initialState);

  const dispatch = React.useCallback(asyncer(baseDispatch, state), []);

  const actions = createActions(dispatch, state);

  return (
    <FyghtContext.Provider value={{ state, ...actions }}>
      {children}
    </FyghtContext.Provider>
  );
};

const useFyghtContext = () => useContext(FyghtContext);

export { FyghtProvider, useFyghtContext };
