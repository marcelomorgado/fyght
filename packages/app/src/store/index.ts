import { FyghtProvider, useFyghtContext } from "./FyghtContext";
export { FyghtProvider, useFyghtContext };
