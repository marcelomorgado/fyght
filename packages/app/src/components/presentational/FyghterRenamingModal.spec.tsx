import * as React from "react";
import { shallow } from "enzyme";
import * as FyghterContext from "../../store";
import { FyghterRenamingModal } from "./FyghterRenamingModal";
import { storeMocks } from "../../testHelpers";

const { myFyghter } = storeMocks;

describe("FyghterRenamingModal", () => {
  test("should render the component", () => {
    const contextValues = { renameMyFyghter: jest.fn(), state: { myFyghter } };

    jest.spyOn(FyghterContext, "useFyghtContext").mockImplementation(() => contextValues);

    expect(shallow(<FyghterRenamingModal />)).toMatchSnapshot();
  });
});
