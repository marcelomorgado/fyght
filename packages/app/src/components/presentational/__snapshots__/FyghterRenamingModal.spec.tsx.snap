// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FyghterRenamingModal should render the component 1`] = `
<div>
  <Button
    block={true}
    ghost={false}
    htmlType="button"
    loading={false}
    onClick={[Function]}
    type="primary"
  >
    Rename
  </Button>
  <FyghterRenamingForm
    name="John"
    onCancel={[Function]}
    onSave={[Function]}
    visible={false}
  />
</div>
`;
